cache/watcher.go

```diff
@@ -20,40 +20,49 @@ import (
 	clientv3 "go.etcd.io/etcd/client/v3"
 )
 
-// watcher holds one client’s buffered stream of events.
+// watcher holds one client’s buffered stream of batches.
 type watcher struct {
-	eventQueue chan *clientv3.Event
+	batchQueue chan *batch
 	keyPred    KeyPredicate
 	stopped    int32
 	done       chan struct{} // closed together with Stop()
 }
 
 func newWatcher(bufSize int, pred KeyPredicate) *watcher {
 	return &watcher{
-		eventQueue: make(chan *clientv3.Event, bufSize),
+		batchQueue: make(chan *batch, bufSize),
 		keyPred:    pred,
 		done:       make(chan struct{}),
 	}
 }
 
-// true  -> event delivered (or filtered/duplicate)
+// true  -> batch delivered (or filtered/duplicate)
 // false -> buffer full (caller should mark watcher “lagging”)
-func (w *watcher) enqueueEvent(event *clientv3.Event) bool {
-	if w.keyPred != nil && !w.keyPred(event.Kv.Key) {
-		return true
+func (w *watcher) enqueueBatch(b *batch) bool {
+	if w.keyPred != nil {
+		filtered := make([]*clientv3.Event, 0, len(b.events))
+		for _, event := range b.events {
+			if w.keyPred(event.Kv.Key) {
+				filtered = append(filtered, event)
+			}
+		}
+		if len(filtered) == 0 {
+			return true
+		}
+		b = &batch{rev: b.rev, events: filtered}
```

Comment:
Enqueue doesn't care about revision in batch, as so it should accept []*clientv3.Events